src/reporting/Market.sol

```diff
@@ -0,0 +1,593 @@
+        if (automatedReporterDisputeBondToken == _shadyBondToken) {
+            return true;
+        }
+        if (limitedReportersDisputeBondToken == _shadyBondToken) {
+            return true;
+        }
+        if (allReportersDisputeBondToken == _shadyBondToken) {
+            return true;
+        }
+        return false;
+    }
+
+    function canBeReportedOn() public constant returns (bool) {
+        // CONSIDER: should we check if migration is necessary here?
+        if (isFinalized()) {
+            return false;
+        }
+        if (!reportingWindow.isReportingActive()) {
+            return false;
+        }
+        return true;
+    }
+
+    function needsMigration() public constant returns (bool) {
+        if (isFinalized()) {
+            return false;
+        }
+        Market _forkingMarket = getBranch().getForkingMarket();
+        if (address(_forkingMarket) == NULL_ADDRESS) {
+            return false;
+        }
+        if (_forkingMarket == this) {
+            return false;
+        }
+        if (block.timestamp < endTime) {
+            return false;
+        }
+        if (automatedReporterAddress != NULL_ADDRESS && block.timestamp < getAutomatedReportDueTimestamp()) {
+            return false;
+        }
+        if (automatedReportReceived && block.timestamp < getAutomatedReportDisputeDueTimestamp()) {
+            return false;
+        }
+        if (automatedReportReceived && address(automatedReporterDisputeBondToken) == NULL_ADDRESS) {
+            return false;
+        }
+        return true;
+    }
+
+    function getAutomatedReportDueTimestamp() public constant returns (uint256) {
+        return endTime + AUTOMATED_REPORTING_DURATION_SECONDS;
+    }
+
+    function getAutomatedReportDisputeDueTimestamp() public constant returns (uint256) {
+        return getAutomatedReportDueTimestamp() + AUTOMATED_REPORTING_DISPUTE_DURATION_SECONDS;
+    }
+
+    function getReportingState() public constant returns (ReportingState) {
+        if (isFinalized()) {
+            return ReportingState.FINALIZED;
+        }
+
+        if (needsMigration()) {
+            return ReportingState.AWAITING_MIGRATION;
+        }
+
+        if (address(allReportersDisputeBondToken) != NULL_ADDRESS) {
+            if (canFinalizeFork()) {
+                return ReportingState.AWAITING_FINALIZATION;
+            }
+            return ReportingState.FORKING;
+        }
+
+        if (address(limitedReportersDisputeBondToken) != NULL_ADDRESS) {
+            if (canFinalizeReporting()) {
+                return ReportingState.AWAITING_FINALIZATION;
+            }
+            if (reportingWindow.isDisputeActive()) {
+                return ReportingState.ALL_DISPUTE;
+            }
+            return ReportingState.ALL_REPORTING;
+        }
+
+        if (reportingWindow.isDisputeActive()) {
+            return ReportingState.LIMITED_DISPUTE;
+        }
+
+        if (canBeReportedOn()) {
```

Comment:
I think this would be more clear (given context) if it just did `if (!reportingWindow.isReportingActive())` instead.  This would align better with the check above this, and also it would avoid the excess `isFinalized` check done in `canBeReportedOn`.